### PR TITLE
wires

### PR DESCRIPTION
**Pull request by ziofil**
**Context:**
We add the `Wires` class for circuit wire operation 

**Description of the Change:**
New Wires class and tests. `Wires` handle `ids` and `indices` for anything related to tensor networks and `CircuitComponent`s.

**Benefits:**
Much simpler implementation and more functionality than previous attempts

**Possible Drawbacks:**
Modes are sorted (so the object needs to account for eventual index reshufflings, not the Wires)

### CODE REVIEW

**Comment by sylviemonet** [comment]
Nice PR combining the index issue in Bargmann and the ids in Fock by using the magic array.

Just have a small question about the "copy" idea in the view. As I understand, we can achieve the view by knowing which row or column that we don't want and then show them, so I haven't fully understood why we need this "copy" in this case.

I've reviewed the first round and together with Sam's PR on wires as well. I'd like to suggest to discuss the combination of the both PRs (such as the jupyter_sphinx, module things).

**Comment by ziofil** [comment]
> Nice PR combining the index issue in Bargmann and the ids in Fock by using the magic array.
> 
> Just have a small question about the "copy" idea in the view. As I understand, we can achieve the view by knowing which row or column that we don't want and then show them, so I haven't fully understood why we need this "copy" in this case.
> 
> I've reviewed the first round and together with Sam's PR on wires as well. I'd like to suggest to discuss the combination of the both PRs (such as the jupyter_sphinx, module things, uuid or random integer, and adjoint of the wires?).

I think that composing them like `wires.ket.output` would not work if ket didn't create a new object. Or perhaps you have figured out a better way?

**Comment by ziofil** [comment]
Also, I've removed adjoint and dual: we don't need them here, just in the circuit components. And the random int is because uuid was giving us an int larger than int64.

**Comment by sylviemonet** [comment]
> Reference

I just noticed that that function is not only related to the index but also return Wires object. Now I agree with this "copy". Thanks.

File: mrmustard/lab_dev/wires.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        all_modes = sorted(set(self.modes) | set(other.modes))
+        for m in all_modes:
+            self_row = self.id_array[self.modes.index(m)] if m in self.modes else np.zeros(4)
+            other_row = other.id_array[other.modes.index(m)] if m in other.modes else np.zeros(4)
+            assert np.all(np.where(self_row > 0) != np.where(other_row > 0)), "duplicate wires!"
+            modes_rows[m] = [s if s > 0 else o for s, o in zip(self_row, other_row)]
+        w = Wires()
+        w._id_array = np.array([modes_rows[m] for m in sorted(modes_rows)])
+        w.mask = np.ones_like(w._id_array)
+        w._modes = sorted(modes_rows)
```
[the remark this answers is not in the document]
Comment (ziofil) [reply mrmustard/lab_dev/wires.py:120]: good idea

**Comment by ziofil** [comment]
Thank you for the comments, Yuan. I added the changes to the other branch (the code that will go in the prototype)

**Pull request closed without merging**